### PR TITLE
ROLA using Signatures Collector

### PR DESCRIPTION
**Pull request by micbakos-rdx**
This PR integrates signing auth with signatures collector. There are some steps to achieve this:

* Deleted the old implementation which was only capable of signing only one entity at a time.
* Implemented a `Signable` version for signing auth:
    * `AuthIntent` became a Signable and it is the request made to signatures collector when signing auth.
    * Since `AuthIntent`'s footprint is small, the `Signable::Payload` is still an `AuthIntent` (No need for `CompiledAuthIntent`. If you feel this is necessary, I could amend.
    * `AuthIntentHash` is the `SignableID`. This is essentially the rola challenge.
    * `SignedAuthIntent` follows the same logic as `SignedIntent` but it also carries the signatures of each owner (`AddressOfAccountOrPersona`), since such information is needed to the hosts. 
* Removed the sample methods from the `Signable` trait and moved them to a separate trait. There was no need to be in the same trait before. It was my lack of experience in rust.
* Introduced `SigningPurpose` into `SignaturesCollector` which can be either `SignTX` or `ROLA`.
* @GhenadieVP According to your latest changes when resolving the signers, I followed the same logic of just ignoring the signers from input that could not be resolved in profile.

### CODE REVIEW

File: crates/sargon-uniffi/src/signing/transaction_sign_request_input.rs [hunk 1]
```diff
@@ -48,3 +61,6 @@ decl_transaction_sign_request_input!(
     signable: Subintent,
     payload: CompiledSubintent
 );
+decl_transaction_sign_request_input!(
+    signable_and_payload: AuthIntent
+);
```
Comment (micbakos-rdx) [inline crates/sargon-uniffi/src/signing/transaction_sign_request_input.rs:66]: Special case for this one. `AuthIntent` can act as a payload too.

File: crates/sargon/src/radix_connect/wallet_interaction/dapp_wallet_interaction/dapp_to_wallet/dapp_metadata/interaction_metadata.rs [hunk 2]
```diff
@@ -37,7 +37,7 @@ impl HasSampleValues for DappToWalletInteractionMetadata {
     fn sample() -> Self {
         Self::new(
             WalletInteractionVersion::sample(),
-            NetworkID::Stokenet,
+            NetworkID::Mainnet,
```
Comment (micbakos-rdx) [inline crates/sargon/src/radix_connect/wallet_interaction/dapp_wallet_interaction/dapp_to_wallet/dapp_metadata/interaction_metadata.rs:40]: These samples were invalid. The dAppDefinigionAddress was in mainnet, but the NetworkID was in Stokenet.

File: crates/sargon/src/signing/authentication/auth_intent.rs [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+        }
+
+        for entity in &entities_to_sign {
+            if entity.network_id() != metadata.network_id {
+                return Err(CommonError::NetworkDiscrepancy {
+                    expected: metadata.network_id,
+                    actual: entity.network_id(),
+                });
+            }
+        }
```
Comment (micbakos-rdx) [inline crates/sargon/src/signing/authentication/auth_intent.rs:50]: Added some validation when generating an `AuthIntent`. I think this is what we need.

File: crates/sargon/src/signing/signables/signable_auth_intent.rs [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+    type Signed = SignedAuthIntent;
+
+    fn entities_requiring_signing(
+        &self,
+        profile: &Profile,
+    ) -> Result<IndexSet<AccountOrPersona>> {
+        let entities = self
+            .entities_to_sign
+            .iter()
+            .filter_map(|address| match address {
```
Comment (micbakos-rdx) [inline crates/sargon/src/signing/signables/signable_auth_intent.rs:15]: @GhenadieVP Ignoring the entities which cannot be resolved in profile.

File: crates/sargon-uniffi/src/signing/transaction_sign_request_input.rs [hunk 5]
```diff
@@ -38,6 +38,19 @@ macro_rules! decl_transaction_sign_request_input {
             );
         }
     };
+    (signable_and_payload: $signable:ty) => {
+        paste! {
```
Comment (CyonAlexRDX) [inline crates/sargon-uniffi/src/signing/transaction_sign_request_input.rs:42]: Please use new macro system - preinterpret. See the matrix macro

Comment (CyonAlexRDX) [reply crates/sargon-uniffi/src/signing/transaction_sign_request_input.rs:42]: https://github.com/radixdlt/sargon/blob/main/crates/sargon-uniffi/src/profile/mfa/security_structures/matrices/decl_matrix_macro.rs

Comment (CyonAlexRDX) [reply crates/sargon-uniffi/src/signing/transaction_sign_request_input.rs:42]: Rationale: crate `paste` is no longer maintained and is gonna be deprecated.

preinterpret is also better.

Created by @dhedey so ask in Rust channel for help from him

Comment (CyonAlexRDX) [reply crates/sargon-uniffi/src/signing/transaction_sign_request_input.rs:42]: But also we have bigger fish to fry... feel free to ignore

File: crates/sargon/src/signing/petition_types/petition_for_entity.rs [hunk 6]
```diff
@@ -76,18 +77,33 @@ impl<ID: SignableID> PetitionForEntity<ID> {
...
-                )
-            }
+            EntitySecurityState::Securified { value } => match purpose {
+                SigningPurpose::SignTX { role_kind } => {
+                    let general_role =
+                            GeneralRoleWithHierarchicalDeterministicFactorInstances::try_from(
+                                (value.security_structure.matrix_of_factors, role_kind)
+                            ).unwrap();
+
+                    PetitionForEntity::new_securified(
```
Comment (CyonAlexRDX) [inline crates/sargon/src/signing/petition_types/petition_for_entity.rs:87]: Can you change `PetitionForEntity` -> `Self` (probably my old code)

File: crates/sargon/src/signing/petition_types/petition_for_entity.rs [hunk 7]
```diff
@@ -76,18 +77,33 @@ impl<ID: SignableID> PetitionForEntity<ID> {
...
+                        general_role,
+                    )
+                }
+                SigningPurpose::ROLA => {
+                    let threshold_factors =
+                        PetitionForFactors::<ID>::new_threshold(
+                            vec![value.authentication_signing_factor_instance()],
+                            1,
+                        );
+                    PetitionForEntity::new(
```
Comment (CyonAlexRDX) [inline crates/sargon/src/signing/petition_types/petition_for_entity.rs:99]: `PetitionForEntity` -> `Self`

File: crates/sargon/src/signing/authentication/auth_intent.rs [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+
+impl AuthIntent {
+    /// Creates an `AuthIntent` from the request received from a dApp.
+    ///
+    /// Fails if the `OriginUrl` in metadata is not a valid url or the network id supplied does not
+    /// match the network id of the given dApp definition address or the entities to sign.
+    pub fn new_from_request(
+        challenge_nonce: DappToWalletInteractionAuthChallengeNonce,
+        metadata: DappToWalletInteractionMetadata,
+        entities_to_sign: Vec<AddressOfAccountOrPersona>,
```
Comment (CyonAlexRDX) [inline crates/sargon/src/signing/authentication/auth_intent.rs:31]: prefer `impl IntoIterator<Item = AddressOfAccountOrPersona >

File: crates/sargon/src/system/sargon_os/sargon_os_signing.rs [hunk 9]
```diff
@@ -106,27 +97,25 @@ mod test {
             "https://example.com",
             DappDefinitionAddress::sample(),
         );
+        let auth_intent = AuthIntent::new_from_request(
+            nonce,
+            metadata,
+            vec![AddressOfAccountOrPersona::Account(account.address)],
```
Comment (CyonAlexRDX) [inline crates/sargon/src/system/sargon_os/sargon_os_signing.rs:103]: `vec!` can be removed if: https://github.com/radixdlt/sargon/pull/310/files#r1891586418

File: crates/sargon/src/signing/signables/signable_auth_intent.rs [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+                AddressOfAccountOrPersona::Account(account_address) => profile
```
Comment (CyonAlexRDX) [inline crates/sargon/src/signing/signables/signable_auth_intent.rs:16]: this should be a method on Profile - getting of entities

File: crates/sargon/src/signing/authentication/auth_intent_hash.rs [hunk 11]
```diff
@@ -169,6 +166,21 @@ mod tests {
...
     }
 
+    fn sut(
+        nonce: Exactly32Bytes,
+        metadata: DappToWalletInteractionMetadata,
+    ) -> Result<SUT> {
+        let intent = AuthIntent::new_from_request(
+            DappToWalletInteractionAuthChallengeNonce(nonce),
+            metadata.clone(),
+            vec![AddressOfAccountOrPersona::Account(AccountAddress::random(
```
Comment (CyonAlexRDX) [inline crates/sargon/src/signing/authentication/auth_intent_hash.rs:176]: can drop `vec!` if ctor uses impl intoiterator

**Review by CyonAlexRDX — APPROVED** [review]
Good job!

**Comment by micbakos-rdx** [comment]
Added some tests regarding Rola. I don't understand why coverage thinks that some code is not tested. The tests clearly pass from such places.

**Pull request merged**